contracts/redemption-rate-adapter/src/contract.rs

```diff
@@ -0,0 +1,149 @@
+use cosmwasm_std::{attr, from_json, to_json_binary, Attribute, Decimal, Deps};
+use cosmwasm_std::{Binary, DepsMut, Env, MessageInfo, Response};
+use cw_ownable::{get_ownership, update_ownership};
+use drop_helpers::answer::response;
+use drop_staking_base::error::redepmtion_rate_adapter::{ContractError, ContractResult};
+use drop_staking_base::msg::redepmtion_rate_adapter::{
+    Config, ExecuteMsg, InstantiateMsg, MigrateMsg, QueryMsg, RedemptionRateResponse,
+};
+
+use drop_staking_base::state::redemtion_rate_adapter::CONFIG;
+use neutron_sdk::bindings::msg::NeutronMsg;
+use neutron_sdk::bindings::query::NeutronQuery;
+
+const CONTRACT_NAME: &str = concat!("crates.io:drop-staking__", env!("CARGO_PKG_NAME"));
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), cosmwasm_std::entry_point)]
+pub fn instantiate(
+    deps: DepsMut<NeutronQuery>,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> ContractResult<Response<NeutronMsg>> {
+    cw2::set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    cw_ownable::initialize_owner(deps.storage, deps.api, Some(msg.owner.as_ref()))?;
+
+    let core_contract = deps.api.addr_validate(&msg.core_contract)?;
+    let config = &Config {
+        core_contract: core_contract.to_string(),
+        denom: msg.denom,
+    };
+    CONFIG.save(deps.storage, config)?;
+
+    Ok(response(
+        "instantiate",
+        CONTRACT_NAME,
+        [attr("core_contract", core_contract)],
+    ))
+}
+
+#[cfg_attr(not(feature = "library"), cosmwasm_std::entry_point)]
+pub fn query(deps: Deps<NeutronQuery>, env: Env, msg: QueryMsg) -> ContractResult<Binary> {
+    match msg {
+        QueryMsg::Ownership {} => Ok(to_json_binary(&get_ownership(deps.storage)?)?),
+        QueryMsg::Config {} => query_config(deps, env),
+        QueryMsg::RedemptionRate { denom, .. } => query_redemption_rate(deps, env, denom),
+    }
+}
+
+fn query_config(deps: Deps<NeutronQuery>, _env: Env) -> ContractResult<Binary> {
+    let config = CONFIG.load(deps.storage)?;
+    Ok(to_json_binary(&config)?)
+}
+
+fn query_redemption_rate(
+    deps: Deps<NeutronQuery>,
+    env: Env,
+    denom: String,
+) -> ContractResult<Binary> {
+    let config = CONFIG.load(deps.storage)?;
+    let core_contract = deps.api.addr_validate(&config.core_contract)?;
```

Comment:
No need to validate this address twice

Reply:
This happens because address is stored as a string and not as `cosmwasm_std::Addr`